monitor: add docstrings to pybind interface

d4l3k opened the pull request:
This adds argument names and docstrings so the docs are a lot more understandable.

Test plan:

docs/tests CI should suffice

![Screenshot 2022-01-19 at 16-35-10 torch monitor — PyTorch master documentation](https://user-images.githubusercontent.com/909104/150240882-e69cfa17-e2be-4569-8ced-71979a89b369.png)

edward-io commented on docs/source/monitor.rst:
```diff
@@ -9,6 +9,16 @@ torch.monitor
 PyTorch.
 
+The stat interfaces are designed to be used for tracking high level metrics that
+are periodically logged out to be used for monitoring system performance. Since
```
"logged out" => "logged to external systems" or just "logged"?

~~to be used~~

edward-io commented on docs/source/monitor.rst:
```diff
@@ -9,6 +9,16 @@ torch.monitor
 
+The stat interfaces are designed to be used for tracking high level metrics that
+are periodically logged out to be used for monitoring system performance. Since
+the stats internally aggregate and have a specific window size you can log to
```
internally aggregate => aggregate internally?

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
+      m,
+      "Aggregation",
+      R"DOC(
+        These are the types of aggregations that can be used as part of the
```
"These are types of aggregations that can be used to accumulate stats"

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
+          "MEAN",
+          Aggregation::MEAN,
+          R"DOC(
+            MEAN computes the mean of all the added values.
```
arithmetic mean

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
+          "COUNT",
+          Aggregation::COUNT,
+          R"DOC(
+            COUNT returns the number of added values.
```
total number

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
+          "get",
+          &Stat<double>::get,
+          R"DOC(
+            Returns the current value of the stat primary for testing purposes.
```
, primarily

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
+          &Stat<double>::get,
+          R"DOC(
+            Returns the current value of the stat primary for testing purposes.
+            If the stat has logged this will be zero.
```
If this stat has been logged out, and no additional values have been added

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
+      m,
+      "IntervalStat",
+      R"DOC(
+        IntervalStat is a Stat that logs the stat once every `window_size`
```
is a Stat that logs ~the stat~ once every

also mention that this is in milliseconds?

d4l3k replied:
updated to show the function signature since it takes a timedelta

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
 
-  py::class_<FixedCountStat<double>, Stat<double>>(m, "FixedCountStat")
-      .def(py::init<std::string, std::vector<Aggregation>, int64_t>());
+        The stat will be log as an event on the next add call after the window
```
log => logged, add =>  \`add\`, end => ends

d4l3k replied:
done

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
+      "FixedCountStat",
+      R"DOC(
+        FixedCountStat is a Stat that logs the stat every `window_size` number
+        of add calls. For high performance stats this window size should be
```
add => \`add\`

d4l3k replied:
done

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
+        of add calls. For high performance stats this window size should be
+        fairly large to ensure that the event logging frequency is in the range
+        of 1s to 60s under normal usage. Core stats should error on the side of
+        less frequent.
```
frequent => frequency

d4l3k replied:
done

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
+      m,
+      "Event",
+      R"DOC(
+        Event represents specific typed event to be logged. This can represent
```
represents specific => represents a specific

d4l3k replied:
done

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -85,31 +85,128 @@ void initMonitorBindings(PyObject* module) {
+      "Event",
+      R"DOC(
+        Event represents specific typed event to be logged. This can represent
+        high data points such as loss or accuracy per epoch or more low level
```
high => high-level, low level => low-level

d4l3k replied:
done

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -137,17 +244,35 @@ void initMonitorBindings(PyObject* module) {
-    return handler;
-  });
+      eventHandlerClass(m, "EventHandler", R"DOC(
+        EventHandler is a wrapper type returned by ``register_event_handler``
```
I missed this earlier, but this may be confusing, since we register event_handlers with a `handler`. Maybe EventHandlerIdentifier?

d4l3k replied:
renamed

edward-io reviewed (commented):
some wording nits

edward-io commented on torch/csrc/monitor/python_init.cpp:
```diff
@@ -122,32 +230,83 @@ void initMonitorBindings(PyObject* module) {
+      },
+      py::arg("handler"),
+      R"DOC(
+        unregister_event_handler unregisters the ``EventHandler`` returned after
```
EventHandler => EventHandlerHandle

edward-io reviewed (approved):
Looks good, one nit

d4l3k commented:
Failure seems to be pre-existing in trunk
```
FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/Version.cpp.o 
/usr/bin/c++  -DADD_BREAKPAD_SIGNAL_HANDLER -DAT_PER_OPERATOR_HEADERS -DCPUINFO_SUPPORTED_PLATFORM=1 -DFMT_HEADER_ONLY=1 -DFXDIV_USE_INLINE_ASSEMBLY=0 -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DNNP_CONVOLUTION_ONLY=0 -DNNP_INFERENCE_ONLY=0 -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DUSE_C10D_GLOO -DUSE_DISTRIBUTED -DUSE_EXTERNAL_MZCRC -DUSE_RPC -DUSE_TENSORPIPE -D_FILE_OFFSET_BITS=64 -Dtorch_cpu_EXPORTS -Iaten/src -I../aten/src -I. -I../ -I../cmake/../third_party/benchmark/include -I../third_party/onnx -Ithird_party/onnx -I../third_party/foxi -Ithird_party/foxi -I../torch/csrc/api -I../torch/csrc/api/include -I../caffe2/aten/src/TH -Icaffe2/aten/src/TH -Icaffe2/aten/src -I../caffe2/../third_party -I../caffe2/../third_party/breakpad/src -Icaffe2/../aten/src -Icaffe2/../aten/src/ATen -I../torch/csrc -I../third_party/miniz-2.0.8 -I../third_party/kineto/libkineto/include -I../third_party/kineto/libkineto/src -I../torch/csrc/distributed -I../aten/../third_party/catch/single_include -I../aten/src/ATen/.. -Icaffe2/aten/src/ATen -I../third_party/FXdiv/include -I../c10/.. -I../third_party/pthreadpool/include -I../third_party/cpuinfo/include -I../third_party/QNNPACK/include -I../aten/src/ATen/native/quantized/cpu/qnnpack/include -I../aten/src/ATen/native/quantized/cpu/qnnpack/src -I../third_party/cpuinfo/deps/clog/include -I../third_party/NNPACK/include -I../third_party/fbgemm/include -I../third_party/fbgemm -I../third_party/fbgemm/third_party/asmjit/src -I../third_party/FP16/include -I../third_party/tensorpipe -Ithird_party/tensorpipe -I../third_party/tensorpipe/third_party/libnop/include -I../third_party/fmt/include -Ithird_party/ideep/mkl-dnn/third_party/oneDNN/include -I../third_party/ideep/mkl-dnn/third_party/oneDNN/src/../include -I../third_party/flatbuffers/include -isystem third_party/gloo -isystem ../cmake/../third_party/gloo -isystem ../cmake/../third_party/googletest/googlemock/include -isystem ../cmake/../third_party/googletest/googletest/include -isystem ../third_party/protobuf/src -isystem /opt/intel/include -isystem ../third_party/gemmlowp -isystem ../third_party/neon2sse -isystem ../third_party/XNNPACK/include -isystem ../third_party -isystem ../cmake/../third_party/eigen -isystem /opt/conda/include/python3.8 -isystem /opt/conda/lib/python3.8/site-packages/numpy/core/include -isystem ../cmake/../third_party/pybind11/include -isystem ../third_party/ideep/mkl-dnn/third_party/oneDNN/include -isystem ../third_party/ideep/include -isystem include -Wno-deprecated-declarations -D_GLIBCXX_USE_CXX11_ABI=1 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -DHAVE_AVX2_CPU_DEFINITION -O3 -DNDEBUG -DNDEBUG -fPIC   -DCAFFE2_USE_GLOO -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-type-limits -Wno-array-bounds -Wno-sign-compare -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -O2 -g -fopenmp -DCAFFE2_BUILD_MAIN_LIB -pthread -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/Version.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/Version.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/Version.cpp.o -c ../aten/src/ATen/Version.cpp
In file included from ../caffe2/core/common.h:27:0,
                 from ../aten/src/ATen/Version.cpp:13:
./caffe2/core/macros.h:12:30: error: ‘ciflow’ was not declared in this scope
 #define CAFFE2_VERSION_MINOR ciflow/macos/71481+cpu
```

d4l3k committed:
monitor: add docstrings to pybind interface (#71481)

Summary:
This adds argument names and docstrings so the docs are a lot more understandable.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/71481

Test Plan:
docs/tests CI should suffice

![Screenshot 2022-01-19 at 16-35-10 torch monitor — PyTorch master documentation](https://user-images.githubusercontent.com/909104/150240882-e69cfa17-e2be-4569-8ced-71979a89b369.png)

Reviewed By: edward-io

Differential Revision: D33661255

Pulled By: d4l3k

fbshipit-source-id: af744a4b8847da5bc3c7f4cedff915b524ca0c39
